ccswbs/gus
# Add Widget block and improve backend UI

# Summary of changes
Add a custom block type capable of using existing widgets which in turn permits the creation of more complex reusable content. Improve the Content Hub backend to make filtering, finding, and tagging content faster and easier.

## Frontend
- Refactor `blockWidget.js` to add new `Widget block` custom block type
- Add necessary widget block schema to `gatsby-node.js`
- Add Block component import to `widget.js` (previously it only worked in `sectionWidget.js`)

### Incidental changes
- Upgrade npm packages to latest versions
- Enable [Fastbuilds](https://www.gatsbyjs.com/plugins/gatsby-source-drupal/#fastbuilds)
- Fix keys and null checks in various component files

## Backend
- Add new custom block type Widget block
- Make Widget block available to field_widgets and field_section_widgets
- Add [bulk_edit_terms module](https://www.drupal.org/project/bulk_edit_terms)
- Add [safedelete module](https://www.drupal.org/project/safedelete)
- Enable file_delete_ui
- Enable views_bulk_operations
- Give required Fastbuilds permission to Gatsby Drupal users
- Default Content view: add search filters (tags, author) and URL alias field
- Default Files view: add search filters and delete option
- Default Media view: add search filters, enable bulk operations

# Test Plan

Gatsby Cloud side: https://tqtest62584.gatsbyjs.io/
Drupal multidev: https://widgetblock-chug.pantheonsite.io/

- Verify all widgets work as expected on https://tqtest62584.gatsbyjs.io/block-test
- Verify the widget block `Admission Requirements BA` works on https://tqtest62584.gatsbyjs.io/programs/studio-art and https://tqtest62584.gatsbyjs.io/programs/bachelor-of-arts
- Review the widget block `Admission Requirements BA` on https://widgetblock-chug.pantheonsite.io/block/27
- Review the widget block `Widget block demo` on https://widgetblock-chug.pantheonsite.io/block/28
- Experiment with creating new widget blocks
- Go to https://widgetblock-chug.pantheonsite.io/admin/content and use the `Update terms` option from the `Actions` dropdown to add tags to selected pieces of content (note: the tag has to already exist)
- Go to https://widgetblock-chug.pantheonsite.io/admin/content/files and click the `Delete` button for an image file. You should see a message telling you the file is being used in a Media node with a link to that Media node.
- Use the new `Use count` search filter to display files with a use count of NULL
- Click `Delete` and verify that you get a "none were found" message with respect to other nodes using the file
- In the Files list, find `Giannina Descalzi 250.jpg` (you can search by Filename or use the new `File URL` field, which will search for words in the file's URL like "jpg" or "inline" or another folder name)
- Click `Delete` and verify you get a message telling you which node is using this file (which *isn't* connected to a Media item). You should also see a link to the node in question (node: the use count includes node revisions)

## Known issues
- The SafeDelete module's `List of orphaned nodes` feature keeps timing out on our site so it's been disabled (the option still displays but no node types have been selected for it to check)
- At this time, SafeDelete can only check the body field for node links, not paragraph types. It does, however, let you know exactly where a file is being used
- It is not possible to filter by URL alias in the content view at this time
